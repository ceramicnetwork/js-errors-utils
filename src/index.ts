/**
 * ```sh
 * npm install errors-utils
 * ```
 *
 * ## Examples
 *
 * ### Library errors with metadata
 *
 * ```ts
 * import { assertAs, createNamespaceError } from 'errors-utils'
 * import { name, version } from '../package.json' // use metadata from package.json
 *
 * const LibError = createNamespaceError('LIB', { package: name, version })
 *
 * const input = 'any'
 * try {
 *   assertAs(typeof input === 'string', LibError, 120, 'Input must be string')
 *   assertAs(input === 'foo', LibError, 123, 'Input must be foo')
 * } catch (error) {
 *   console.log(error.toString()) // '[LIB123] Input must be foo'
 *   throw new LibError(10, 'Input validation failed', error) // Wrap thrown error
 * }
 * ```
 *
 * ### Custom assertions
 *
 * ```ts
 * import { StackError, assertAs, createNamespaceError } from 'errors-utils'
 *
 * function createAssert(ErrorClass: typeof StackError) {
 *   // Our assert function will use the provided Error class and default message
 *   function assert(condition: boolean, code: string | number, message = 'Assertion failed') {
 *     return assertAs(condition, ErrorClass, code, message)
 *   }
 *
 *   assert.equal = (a, b, code = 11, msg = `${a} must be equal to ${b}`) => {
 *     return assert(a === b, code, msg)
 *   }
 *   assert.notEqual = (a, b, code = 12, msg = `${a} must not be equal to ${b}`) => {
 *     return assert(a !== b, code, msg)
 *   }
 *   // ...
 *
 *   return assert
 * }
 *
 * const LibError = createNamespaceError('LIB')
 * const assert = createAssert(LibError)
 * assert.equal(a, b)
 * ```
 *
 * ### Error classes extensions
 *
 * ```ts
 * import { assertAs, createNamespaceError } from 'errors-utils'
 *
 * // ProtocolError is used for clients interactions
 *
 * class ProtocolError extends createNamespaceError('PTL') {
 *   toAPI(response) { ... }
 * }
 *
 * const PROTOCOL_VERSION = 2
 *
 * function assertProtocolVersion(version: number) {
 *   return assertAs(version === PROTOCOL_VERSION, ProtocolError, 1, `Invalid protocol version: expected ${PROTOCOL_VERSION}, got ${version}`)
 * }
 *
 * function handleAPICall(request, response) {
 *   try {
 *     assertProtocolVersion(request.body.version)
 *   } catch (error) {
 *     if (error instanceof ProtocolError) {
 *       return error.toAPI(response)
 *     }
 *   }
 * }
 *
 * // InternalError is used for platform interactions
 *
 * class InternalError extends createNamespaceError('INT') {
 *   // Attach logger to instance
 *   logger: Logger = myLogger
 *
 *   log(level = 'critical') {
 *     this.logger.log({ level, code: this.code, message: this.message })
 *   }
 * }
 *
 * function assertValidService(service) {
 *   return assertAs(service instanceof Service, InternalError, 123, 'Invalid service provided')
 * }
 *
 * function checkConfig(config) {
 *   try {
 *     assertValidService(config.myService)
 *     ...
 *   } catch (err) {
 *     if (error instanceof InternalError) {
 *       error.log()
 *     }
 *   }
 * }
 * ```
 */

export type StackErrorJSON = {
  code: string
  message: string
  metadata: Record<string, unknown>
  name: string
  stack: Array<StackErrorJSON>
}

export class StackError extends Error {
  /**
   * Casts an `Error` to a `StackError`, using the given `code`. Calling this function with an
   * instance of `StackError` will return the input unchanged.
   */
  static from(error: Error, code = 'SE0'): StackError {
    if (error instanceof StackError) {
      return error
    }
    const se = new StackError(code, error.message)
    se.stack = error.stack
    return se
  }

  static fromJSON(json: StackErrorJSON): StackError {
    const error = new StackError(json.code, json.message)
    error.errorStack = (json.stack ?? []).reduceRight((stack, e) => {
      const err = StackError.fromJSON(e)
      err.errorStack = stack
      return [err, ...stack]
    }, [] as Array<StackError>)
    error.metadata = json.metadata ?? {}
    error.name = json.name ?? 'StackError'
    return error
  }

  code: string
  errorStack: Array<StackError>
  metadata: Record<string, unknown> = {}
  name = 'StackError'

  constructor(code: string, message: string, wrapError?: Error) {
    super(message)
    Object.setPrototypeOf(this, StackError.prototype)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, StackError)
    }

    this.code = code
    this.errorStack = wrapError ? StackError.from(wrapError).toErrorStack() : []
  }

  toErrorStack(): Array<StackError> {
    return [this, ...this.errorStack]
  }

  /**
   * Serializes the error to JSON. By default the `errorStack` is included on a single level,
   * setting the `withStack` argument to `false` will serialize the error only, discarding the
   * stack.
   */
  toJSON(withStack = true): StackErrorJSON {
    return {
      code: this.code,
      message: this.message,
      metadata: this.metadata,
      name: this.name,
      stack: withStack ? this.errorStack.map((e) => e.toJSON(false)) : [],
    }
  }

  toString(): string {
    return `[${this.code}] ${this.message}`
  }
}

/**
 * Factory for an Error class extending `StackError` with a given `namespace` and optional
 * `metadata`.
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function createNamespaceError(namespace: string, metadata: Record<string, unknown> = {}) {
  return class NamespaceError extends StackError {
    constructor(code: string | number, message: string, wrapError?: Error) {
      super(`${namespace}${code}`, message, wrapError)
      Object.setPrototypeOf(this, NamespaceError.prototype)
      if (Error.captureStackTrace) {
        Error.captureStackTrace(this, NamespaceError)
      }

      this.metadata = metadata
    }
  }
}

/**
 * Asserts the given `condition` is true or throws an Error with the given `message`.
 */
export function assert(condition: boolean, message = 'Assertion failed'): asserts condition {
  if (!condition) {
    throw new Error(message)
  }
}

/**
 * Asserts the given `condition` is true or throws an Error using the given `ErrorClass` and
 * associated arguments.
 */
export function assertAs<T extends typeof StackError>(
  condition: boolean,
  ErrorClass: T,
  ...args: ConstructorParameters<T>
): asserts condition {
  if (!condition) {
    // @ts-ignore args
    throw new ErrorClass(...args)
  }
}
