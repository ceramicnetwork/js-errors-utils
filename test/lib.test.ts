import { StackError, assert, assertAs, createNamespaceError } from '../src'

describe('lib', () => {
  describe('StackError', () => {
    test('extends Error', () => {
      expect(new StackError('TEST0', 'test')).toBeInstanceOf(Error)
    })

    test('has a code property', () => {
      const error = new StackError('TEST0', 'test')
      expect(error.code).toBe('TEST0')
    })

    test('has an errorStack property', () => {
      const first = new StackError('TEST1', 'first')
      const second = new StackError('TEST2', 'second', first)
      const third = new StackError('TEST3', 'third', second)
      expect(third.errorStack).toEqual([second, first])
    })

    test('has a metadata property', () => {
      const error = new StackError('TEST0', 'test')
      error.metadata = { test: 'meta' }
      expect(error.metadata.test).toBe('meta')
    })

    test('toString() formats the error', () => {
      const error = new StackError('TEST0', 'test')
      expect(error.toString()).toBe('[TEST0] test')
    })

    test('toStack() returns the errors as a list', () => {
      const first = new StackError('TEST1', 'first')
      const second = new StackError('TEST2', 'second', first)
      const third = new StackError('TEST3', 'third', second)
      expect(third.toErrorStack()).toEqual([third, second, first])
    })

    test('can be converted to and from JSON', () => {
      const first = new StackError('TEST1', 'first')
      const second = new StackError('TEST2', 'second', first)
      const third = new StackError('TEST3', 'third', second)
      const clone = StackError.fromJSON(third.toJSON())
      expect(clone).toBeInstanceOf(StackError)
      expect(clone).toEqual(third)
    })
  })

  describe('assert', () => {
    test('throws if the assertion fails', () => {
      expect(() => assert(false, 'should be true')).toThrow('should be true')
    })

    test('throws with a default error message', () => {
      expect(() => assert(false)).toThrow('Assertion failed')
    })

    test('does not throw if the assertion succeeds', () => {
      expect(assert(true)).toBeUndefined()
    })
  })

  describe('assertAs', () => {
    test('throws if the assertion fails', () => {
      expect(() => assertAs(false, StackError, 'TEST', 'should be true')).toThrow(StackError)
      expect(() => assertAs(false, StackError, 'TEST', 'should be true')).toThrow('should be true')
    })

    test('does not throw if the assertion succeeds', () => {
      expect(assertAs(true, StackError, 'TEST', 'should be true')).toBeUndefined()
    })
  })

  describe('createNamespaceError', () => {
    const meta = { foo: 'bar' }
    const TestError = createNamespaceError('TEST', meta)
    const error = new TestError(1, 'test')

    test('extends Error and StackError', () => {
      expect(error).toBeInstanceOf(Error)
      expect(error).toBeInstanceOf(StackError)
    })

    test('injects the namespace and metadata', () => {
      expect(error.code).toBe('TEST1')
      expect(error.metadata).toBe(meta)
    })

    test('can be used with assertAs', () => {
      expect(() => assertAs(false, TestError, 1, 'should be true')).toThrow(TestError)
    })
  })

  test('example code works', () => {
    const LibError = createNamespaceError('LIB', { package: 'my-lib', version: '0.1.0' })
    const input = 'any'

    let output
    try {
      assertAs(typeof input === 'string', LibError, 120, 'Input must be string')
      // @ts-ignore
      assertAs(input === 'foo', LibError, 123, 'Input must be foo')
    } catch (error) {
      const wrapped = new LibError(10, 'Input validation failed', error)
      output = wrapped.toJSON()
    }

    expect(output).toMatchSnapshot()
  })

  test('cross-namespaces works', () => {
    const LibAError = createNamespaceError('LIBA', { package: 'lib-a', version: '0.1.0' })

    function validate(input: string): void {
      assertAs(typeof input === 'string', LibAError, 120, 'Input must be string')
      // @ts-ignore
      assertAs(input === 'foo', LibAError, 123, 'Input must be foo')
    }

    const LibBError = createNamespaceError('LIBB', { package: 'lib-b', version: '0.2.0' })

    let output
    try {
      validate('any')
    } catch (error) {
      const wrapped = new LibBError(10, 'Input validation failed', error)
      output = wrapped.toJSON()
    }

    expect(output).toMatchSnapshot()
  })
})
