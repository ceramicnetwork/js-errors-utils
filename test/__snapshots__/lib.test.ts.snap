// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib cross-namespaces works 1`] = `
Object {
  "code": "LIBB10",
  "message": "Input validation failed",
  "metadata": Object {
    "package": "lib-b",
    "version": "0.2.0",
  },
  "name": "StackError",
  "stack": Array [
    Object {
      "code": "LIBA123",
      "message": "Input must be foo",
      "metadata": Object {
        "package": "lib-a",
        "version": "0.1.0",
      },
      "name": "StackError",
      "stack": Array [],
    },
  ],
}
`;

exports[`lib example code works 1`] = `
Object {
  "code": "LIB10",
  "message": "Input validation failed",
  "metadata": Object {
    "package": "my-lib",
    "version": "0.1.0",
  },
  "name": "StackError",
  "stack": Array [
    Object {
      "code": "LIB123",
      "message": "Input must be foo",
      "metadata": Object {
        "package": "my-lib",
        "version": "0.1.0",
      },
      "name": "StackError",
      "stack": Array [],
    },
  ],
}
`;
